Indexed Entities

iethree opened the pull request:
Adds indexed entities.

In a model's metadata editor, string columns have a new option to index them

![image](https://user-images.githubusercontent.com/6377293/236071535-171262e0-7db2-4d7f-870f-a0323b17e91d.png)


This will index the model, saving the row's primary key along with the value of the column to be indexed. You can now search for those terms in the general search bar:
![image](https://user-images.githubusercontent.com/6377293/236073082-a51eba9e-dd27-4115-be4a-5c626d88c64d.png)

Choosing that search result will take you to that model and the details view of that record:

<img width="1050" alt="image" src="https://user-images.githubusercontent.com/6377293/236071718-3b6980df-3a49-42eb-a804-baa2878ff648.png">

### Data model:

We have a model index table to hold the index:
```sql
# select * from model_index;
-[ RECORD 1 ]---+------------------------------
id              | 287
model_id        | 330
pk_ref          | ["field",62,null]
value_ref       | ["field",65,null]
generation      | 1
schedule        | 0 0 16 * * ? *
state           | indexed
state_change_at | 2023-05-03 18:17:05.642807-05
error           | [null]
created_at      | 2023-05-03 18:17:05.376542-05
creator_id      | 5
```

and a model_index_value table to hold the values:

```sql
# select * from model_index_value limit 6;
 model_index_id | model_pk |            name             | generation
----------------+----------+-----------------------------+------------
            287 |      150 | Synergistic Wool Coat       |          1
            287 |       13 | Synergistic Steel Chair     |          1
            287 |      118 | Synergistic Rubber Shoes    |          1
            287 |       24 | Synergistic Marble Keyboard |          1
            287 |       69 | Synergistic Leather Coat    |          1
            287 |        3 | Synergistic Granite Chair   |          1
```

A task is scheduled once a day to refresh the values.

#### Generation

The generation field is used to efficiently diff the sets for values no longer in the dataset. There's a unique constraint on `(model_pk, model_index_id)` so that we can upsert the name based on the pk and index. New inserts and updates set the generation to the next generation and then we delete all values for that model index with a lower generation. On h2 we just do the simple version of deleting existing keys and inserting all new ones.

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/metabase/metabase/30487)
<!-- Reviewable:end -->

dpsutton committed:
indexed entities schema

dpsutton committed:
schema

dpsutton committed:
endpoint with create and delete

dpsutton committed:
hooked up tasks

dpsutton committed:
first tests and hooking it all together

dpsutton committed:
first tests and hooking it all together

dpsutton committed:
some e2e tests

dpsutton committed:
search

dpsutton committed:
search

had to rename value -> name. The search stuff works well for stuff that
looks the same. And previously was renaming `:value` to name in the
search result, but it uses the searchable-columns-for-model for which
values to score on and wasn't finding :value. So easier to just let name
flow through.

dpsutton committed:
include model_pk in results

dpsutton committed:
send pk back as id, include pk_ref

dpsutton committed:
rename fk constraint (had copied it)

dpsutton committed:
populate indexed values with name not value

dpsutton committed:
GET http://localhost:3000/api/model_index?model_id=135

dpsutton committed:
populate indexed values on post

done here for ease of demo. followup work will schedule the task
immediately, but for now, do it on thread and return info

dpsutton committed:
test search results

dpsutton committed:
fix t2 delete syntax on h2

dpsutton committed:
fix h2

dpsutton committed:
tests work on h2 and postgres app dbs

dpsutton committed:
Fix after insert method

after method lets you change the data. and i was returning the results
of the refresh job as the model-index. not great

dpsutton committed:
unify some cross-db logic for inserting model index values

dpsutton committed:
Extract shared logic in populating indexed values; mysql support

dpsutton committed:
clean ns

dpsutton committed:
I was the bug (endpoint returns the item now)

dpsutton committed:
fix delete to check perms on model not model-index

dpsutton committed:
fix tests

dpsutton committed:
Fix tests

dpsutton committed:
ignore unused private var

dpsutton committed:
Add search tests

dpsutton committed:
remove tap>

dpsutton committed:
Tests for simple mbql, joined mbql, native

dpsutton committed:
shutdown scheduler after finishing

iethree committed:
Entity Search + Indexing Frontend (#30082)

* add model index types and mocks

* add integer type check

* add new entities for model indexes

* expose model index toggle in model metadata

* add search description for indexed entities

* add an error boundary to the app bar

* e2e test entity indexes

* update tests and types

dpsutton committed:
Renumber migrations

i forgot to claim them and i am a bad person

dpsutton committed:
Restore changes lost in the rebase

dpsutton committed:
add sync task to prevent errors in temp

without this you get errors when the temp db is created as it wants to
set the sync and analyze jobs but those jobs haven't been started.

dpsutton committed:
Restore and test GET model-index/:model_id

dpsutton committed:
clean up api: get by model id or model-index-id

iethree committed:
update e2e tests

dpsutton committed:
ensure correct types on id-ref and value-ref

dpsutton committed:
simplify lookup

dpsutton committed:
More extensive testing

iethree committed:
update types

dpsutton committed:
reorder migrations

dpsutton committed:
fix tests

dpsutton committed:
empty to trigger CI

iethree committed:
update types

dpsutton committed:
Bump clj-kondo

old version reports

/work/src/metabase/models/model_index.clj:27:6: error: #'metabase.models.interface/add-created-at-timestamp is private

on source:

```clojure
(t2/define-before-insert ::created-at-timestamp
  [instance]
  #_{:clj-kondo/ignore [:private-call]} ;; <- ignores this override
  (#'mi/add-created-at-timestamp instance))
```

dpsutton committed:
Move task assertions to model namespace

the task system is initialized in the repl so these all work
locally. But in CI that's not necessarily the case. And rather than
mocking the task system again I just leave it in the hands of the other
namespace.

dpsutton committed:
Don't serialize ModelIndex and ModelIndexValue

seems like it is a setting on an individual instance. No need to have
that go across boundaries

dpsutton committed:
Just test this on h2.

chasing names across the different dbs is maddening. and the underlying
db doesn't really matter

qnkhuat commented on resources/migrations/000_migrations.yaml:
```diff
@@ -14511,6 +14511,151 @@ databaseChangeLog:
+                  defaultValueComputed: current_timestamp
+                  constraints:
+                    nullable: false
+              - column:
```
should we have an index for this? -- (PK does not automatically indexed)

dpsutton replied:
good call out. I'm gonna added indexes to model_index.model_id and model_index_value.model_index_id

qnkhuat commented on resources/migrations/000_migrations.yaml:
```diff
@@ -14511,6 +14511,151 @@ databaseChangeLog:
+            remarks: 'Used to keep track of the values indexed in a model'
+            columns:
+              - column:
+                  name: model_index_id
```
we need to create an index for this column

dpsutton replied:
good call out. I'm gonna added indexes to `model_index.model_id` and `model_index_value.model_index_id`

qnkhuat commented on src/metabase/task/index_values.clj:
```diff
@@ -0,0 +1,113 @@
+     (u/format-color :green (trs "Scheduling indexing for model: {0}" (:model_id model-index))))
+    (try (task/add-trigger! trigger)
+         (catch ObjectAlreadyExistsException _e
+           (log/info (u/format-color :green (trs "Persistence already present for model: {0}"
```
```suggestion
           (log/info (u/format-color :green (trs "Persistence already present for model: {0}"
```
this error message seems off? did you mean the model index already present?

dpsutton replied:
copy/paste error. thanks

qnkhuat commented on test/metabase/api/model_index_test.clj:
```diff
@@ -0,0 +1,87 @@
+    (let [query     (mt/mbql-query products)
+          pk_ref    (mt/$ids $products.id)
+          value_ref (mt/$ids $products.title)]
+      (mt/with-temp* [Card [model {:dataset         true
+                                   :name            "model index test"
+                                   :dataset_query   query
+                                   :result_metadata (result-metadata-for-query query)}]]
```
```suggestion
      (mt/with-temp* [Card [model (merge (mt/card-with-source-metadata-for-query query)
       {:dataset true
        :name    "model index test"})]]
```

this does the same thing

qnkhuat replied:
also, I think we should start using `t2.with-temp/with-temp` to make the transition easier

qnkhuat commented on test/metabase/models/model_index_test.clj:
```diff
@@ -0,0 +1,216 @@
+        (mt/with-temp* [Card [model {:dataset         true
+                                     :name            "model index test"
+                                     :dataset_query   query
+                                     :result_metadata (result-metadata-for-query query)}]]
```
same with https://github.com/metabase/metabase/pull/30487/files?file-filters%5B%5D=.clj&file-filters%5B%5D=.yaml&file-filters%5B%5D=.yml&show-viewed-files=true#r1187170127

qnkhuat commented on src/metabase/models/model_index.clj:
```diff
@@ -0,0 +1,150 @@
+(models/defmodel ModelIndexValue :model_index_value)
+
+
+;; switch to the one in mi after #29513 lands
+(t2/define-before-insert ::created-at-timestamp
+  [instance]
+  #_{:clj-kondo/ignore [:private-call]}
+  (#'mi/add-created-at-timestamp instance))
+
+(derive ModelIndex ::created-at-timestamp)
```
#29513 landed, this can be 

```clojure
(doto ModelIndex
  (derive :hook/timestamped?)))
```
now.

dpsutton replied:
i want `:hook/created-at-timestamped?` but yes you are right.

qnkhuat commented on src/metabase/models/model_index.clj:
```diff
@@ -0,0 +1,150 @@
+  (comp #'mbql.normalize/canonicalize-mbql-clauses
+        #'mbql.normalize/normalize-tokens))
+
+(t2/deftransforms ModelIndex
```
wdyt if we move this transformation to `model.interface` like this 

https://github.com/metabase/metabase/blob/9def91c91ff6427b76279f1a39ccdf2d4c4a0da7/src/metabase/models/interface.clj#L383-L386

Probably be useful in the future too.

qnkhuat commented on src/metabase/models/model_index.clj:
```diff
@@ -0,0 +1,150 @@
+(defn- fetch-values
+  [model-index]
+  (let [model (t2/select-one Card :id (:model_id model-index))]
+    (try [nil (->> (qp/process-query
```
```suggestion
    (try [(->> (qp/process-query
```

qnkhuat commented on src/metabase/api/model_index.clj:
```diff
@@ -0,0 +1,81 @@
+  {model_id  ms/PositiveInt
+   pk_ref    any?
+   value_ref any?}
+  (let [model      (api/read-check Card model_id)
```
```suggestion
  (api/check-superuser)
  (let [model      (api/read-check Card model_id)
```

Per the engineering sketch, we should limit this to admins only. We also need this for `DELETE`

qnkhuat commented on src/metabase/api/model_index.clj:
```diff
@@ -0,0 +1,81 @@
+(defn- ensure-type
+  "Ensure that the ref exists and is of type required for indexing."
+  [t ref metadata]
+  (if-let [field (some (fn [f] (when ((comp #{(model-index/normalize-field-ref ref)} :field_ref) f)
```
we have `qp.util/field->field-info` to find `field` in `result_metadata` given a `field_ref`,  there are edge cases where doing a simple equality check is not sufficient( I.e  aggregated fields or field_ref reference by name in native queries)

dpsutton replied:
I'll look into this, but i think we actually do want equality on the field ref. The ref is taken from the metadata so we don't need to do anything fancy to recognize it I think.

qnkhuat replied:
that function actually tries doing an equality check first.

Then it'll fall back to check pair `name, base-type` to cover some edge cases for aggregation + column with id is a string.

qnkhuat commented on src/metabase/api/model_index.clj:
```diff
@@ -0,0 +1,81 @@
+    (when-not (seq metadata)
+      (throw (ex-info (tru "Model has no metadata. Cannot index")
+                      {:model-id model_id})))
+    (ensure-type :type/PK pk_ref metadata)
```
shouldn't we check pk must have base-type is integer too?

dpsutton replied:
yes good point.

qnkhuat commented on src/metabase/task/index_values.clj:
```diff
@@ -0,0 +1,113 @@
+(declare model-index-trigger-key)
+
+(defn- refresh-index!
+  "Function "
```
missing docstring?

dpsutton commented on resources/migrations/000_migrations.yaml:
```diff
@@ -14511,6 +14511,151 @@ databaseChangeLog:
+                  remarks: 'The label to display identifying the indexed value.'
+                  constraints:
+                    nullable: false
+              - column:
```
we need to keep track of the current generation so we know which is the "next" generation

dpsutton committed:
indexes on model_index tables

- `model_index.model_id` for searching by model_id
- `model_index_value.model_index_id` getting values for a particular
  index

dpsutton committed:
copy/paste error in indexing error message

dpsutton committed:
`mt/with-temp` -> `t2.with-temp/with-temp`

dpsutton committed:
use `:hook/created-at-timestamped?`

dpsutton committed:
move field-ref normalization into models.interface

dpsutton committed:
nit: alignment

dpsutton committed:
Ensure write access to underlying model for create/delete

dpsutton committed:
Assert more about pk/value refs and better error messages

dpsutton committed:
Don't search indexed entities when user is sandboxed

Adds a `= 1 0` clause to indexed-entity search if the user is
sandboxed.

Also, got tired of the pattern:

```clojure
(if-let [segmented-user? (resolve 'metabase-enterprise.sandbox.api.util/segmented-user?)]
  (if (segmented-user?)
    :sandboxed-user
    :not-sandboxed-user)
  :not-sandboxed-user)
```

and now we have a tasty lil ole'

```clojure
(defenterprise segmented-user?
  metabase-enterprise.sandbox.api.util
  []
  false)
```

that you can just

```clojure
(ns foo
  (:require
   [metabase.public-settings.premium-features :as premium-features]))

(if (premium-features/segmented-user?)
  :sanboxed
  :not-sandboxed)
```

dpsutton committed:
redef premium-features/segmented-user?

dpsutton committed:
trigger CI

dpsutton committed:
Alternative GEOJSON.io fix (#30675)

* Alternative GEOJSON.io fix

rather than not testing, just accept that we might get an expected
location back or we might get nil. This actually represents how the
application will behave and will seamlessly work as the service improves
or continues throwing errors:

```shell
❯ curl "https://get.geojs.io/v1/ip/geo.json?ip=8.8.8.8,136.49.173.73,185.233.100.23"
<html>
<head><title>500 Internal Server Error</title></head>
<body>
<center><h1>500 Internal Server Error</h1></center>
<hr><center>openresty</center>
</body>
</html>

❯ curl "https://get.geojs.io/v1/ip/geo.json?ip=8.8.8.8,136.49.173.73"
[{"area_code":"0","organization_name":"GOOGLE","country_code":"US",
  "country_code3":"USA","continent_code":"NA","ip":"8.8.8.8",
  "region":"California","latitude":"34.0544","longitude":"-118.2441",
  "accuracy":5,"timezone":"America\/Los_Angeles","city":"Los Angeles",
  "organization":"AS15169 GOOGLE","asn":15169,"country":"United States"},
 {"area_code":"0","organization_name":"GOOGLE-FIBER","country_code":"US",
 "country_code3":"USA","continent_code":"NA","ip":"136.49.173.73",
 "region":"Texas","latitude":"30.2423","longitude":"-97.7672",
 "accuracy":5,"timezone":"America\/Chicago","city":"Austin",
 "organization":"AS16591 GOOGLE-FIBER","asn":16591,"country":"United States"}]
```

Changes are basically

```clojure
(schema= (s/conditional some? <original-expectation-schema>
                        nil?  (s/eq nil)
         response-from-geojson)
```

* Filter to login error messages

```clojure
(into [] (map (fn [[log-level error message]] [log-level (type error) message]))
      error-messages-captured-in-test)
[[:error
  clojure.lang.ExceptionInfo
  "Error geocoding IP addresses {:url https://get.geojs.io/v1/ip/geo.json?ip=127.0.0.1}"]
 [:error clojure.lang.ExceptionInfo "Authentication endpoint error"]]
 ```

* check timestamps with regex

seems like they fixed the service

```shell
curl "https://get.geojs.io/v1/ip/geo.json?ip=8.8.8.8,136.49.173.73,185.233.100.23"
[{"country":"United States","latitude":"34.0544","longitude":"-118.2441","accuracy":5,"timezone":"America\/Los_Angeles","ip":"8.8.8.8","organization":"AS15169 GOOGLE","country_code3":"USA","asn":15169,"area_code":"0","organization_name":"GOOGLE","country_code":"US","city":"Los Angeles","continent_code":"NA","region":"California"},{"country":"United States","latitude":"30.2423","longitude":"-97.7672","accuracy":5,"timezone":"America\/Chicago","ip":"136.49.173.73","organization":"AS16591 GOOGLE-FIBER","country_code3":"USA","asn":16591,"area_code":"0","organization_name":"GOOGLE-FIBER","country_code":"US","city":"Austin","continent_code":"NA","region":"Texas"},{"country":"France","latitude":"48.8582","longitude":"2.3387","accuracy":500,"timezone":"Europe\/Paris","ip":"185.233.100.23","organization":"AS198985 AQUILENET","country_code3":"FRA","asn":198985,"area_code":"0","organization_name":"AQUILENET","country_code":"FR","continent_code":"EU"}]
```

whereas a few hours ago that returned a 500. And now the timestamps are different

```
;; from schema
"2021-03-18T19:52:41.808482Z"
;; results
"2021-03-18T20:52:41.808482+01:00"
```

kinda sick of dealing with this and don't want to deal with a schema
that matches "near" to a timestamp.

* Restore these

i copied changes over from tim's fix and it included these fixes. But
now the service is fixed and these are the original values from cam's
original fixes

dpsutton committed:
Separate FK reference because of index

ERROR in metabase.db.schema-migrations-test/rollback-test (ChangeLogIterator.java:126)
Uncaught exception, not in assertion.

   liquibase.exception.LiquibaseException: liquibase.exception.RollbackFailedException: liquibase.exception.DatabaseException: (conn=637) Cannot drop index 'idx_model_index_model_id': needed in a foreign key constraint [Failed SQL: (1553) DROP INDEX `idx_model_index_model_id` ON `schema-migrations-test-db-32952`.`model_index`]
   liquibase.exception.RollbackFailedException: liquibase.exception.DatabaseException: (conn=637) Cannot drop index 'idx_model_index_model_id': needed in a foreign key constraint [Failed SQL: (1553) DROP INDEX `idx_model_index_model_id` ON `schema-migrations-test-db-32952`.`model_index`]
   liquibase.exception.DatabaseException: (conn=637) Cannot drop index 'idx_model_index_model_id': needed in a foreign key constraint [Failed SQL: (1553) DROP INDEX `idx_model_index_model_id` ON `schema-migrations-test-db-32952`.`model_index`]
   java.sql.SQLTransientConnectionException: (conn=637) Cannot drop index 'idx_model_index_model_id': needed in a foreign key constraint
     SQLState: "HY000"
    errorCode: 1553

dpsutton committed:
delete cascade on model_index fk to report_card

dpsutton committed:
reorder fk and index migrations

dpsutton committed:
break apart FK from table definition for rollbacks

iethree committed:
Merge branch 'master' into indexed-entities-schema

# Conflicts:
#	frontend/src/metabase-types/api/field.ts

iethree committed:
update types and appease the new lint rules

qnkhuat commented on src/metabase/models/model_index.clj:
```diff
@@ -0,0 +1,134 @@
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;; model lifecycle ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+(models/defmodel ModelIndex :model_index)
+(models/defmodel ModelIndexValue :model_index_value)
```
We should use the toucan2 way to define a model: 

https://github.com/metabase/metabase/blob/8f5f36720ab2ee2a70bfed9d150536b6aca9cda8/src/metabase/models/card.clj#L34-L39


```
(methodical/defmethod t2/table-name :model/ModelIndex [_model] :model_index)
(methodical/defmethod t2/table-name :model/ModelIndexValue [_model] :model_index_value)

(derive :model/ModelIndex :metabase/model)
(derive :model/ModelIndexValue :metabase/model)
```

qnkhuat replied:
a bonus point to start using `:model/ModelIndex` instead of `Model` too.

iethree committed:
Merge branch 'master' into indexed-entities-schema

# Conflicts:
#	frontend/src/metabase-types/store/mocks/state.ts

qnkhuat commented on resources/migrations/000_migrations.yaml:
```diff
@@ -14615,6 +14615,198 @@ databaseChangeLog:
+      id: v47.00-019
+      author: dpsutton
+      comment: Indexed Entities values table
+      validCheckSum: ANY
```
IMO we should remove these `validCheckSum` since this is not merged into master.

qnkhuat commented on enterprise/backend/src/metabase_enterprise/sandbox/api/util.clj:
```diff
@@ -34,9 +35,10 @@
+  :feature :sandboxes
   []
   (boolean
    (when-not *is-superuser?*
```
just thinking out loud here: I wonder whether we can cache this smartly. Search perf is already bad and this makes it even worse for enterprise users.

qnkhuat replied:
we could maybe improve the performance a bit by combine the 2 selects into one?

qnkhuat replied:
not necessary to be in this PR. lmk if you want me to open a PR for this.

dpsutton replied:
i'm open to that improvement, but this only impacts enterprise users with the sandbox feature, whereas previously it would impact anyone with the enterprise namespaces on the classpath. It's true search is slow but that's because it does lots of full table scans. These extra app db queries don't insult me at the moment. But I'm fine with a follow-on issue to speed this up.

qnkhuat commented on test/metabase/api/model_index_test.clj:
```diff
@@ -0,0 +1,88 @@
+                                      {:model_id  (:id model)
+                                       :pk_ref    pk_ref
+                                       :value_ref value_ref}))))
+          (testing "GET by model id"
```
403 tests too?

qnkhuat commented on test/metabase/models/model_index_test.clj:
we should have a test for `add-values` against all the supported DBs, I think we're testing for h2 only.

qnkhuat reviewed (dismissed):
This is very close 👏  
but I'd like to see more tests, 
- perms tests, especially testing sandboxed users can't search indexed entities
- `add-values!` tests for all supported DBMS

metamben commented on src/metabase/task/index_values.clj:
```diff
@@ -0,0 +1,116 @@
+  "States of a model index that are refreshable."
+  #{"indexed" "initial" "error" "overflow"})
+
+(defn should-deindex?
```
Does this have to be public?

metamben commented on src/metabase/task/index_values.clj:
```diff
@@ -0,0 +1,116 @@
+
+(defn should-deindex?
+  "Whether to unindex the the model indexing job.
+  If the model does not exist, the model-index record, or if the model has reverted to a question."
```
I don't understand this sentence ("the model-index record") and it doesn't say anything about the model being archived.

metamben commented on src/metabase/task/index_values.clj:
```diff
@@ -0,0 +1,116 @@
+      (not (:dataset model))
+      (:archived model)))
+
+(declare model-index-trigger-key)
```
We could have the definition here, couldn't we?

metamben commented on test/metabase/api/search_test.clj:
```diff
@@ -474,6 +466,62 @@
+  (testing "Should search indexed entities"
+    (mt/dataset airports
+      (let [query (mt/mbql-query municipality)]
+        (mt/with-temp* [Card [model {:dataset       true
```
Could use `mt/with-temp`.

dpsutton replied:
I don't like `mt/with-temp`. Since if you need to add another temp item the syntax changes.

metamben replied:
> I don't like `mt/with-temp`. Since if you need to add another temp item the syntax changes.

I can understand that. I think the mere presence of `mt/with-temp` is causing some confusion. It would be better if it were private.

dpsutton committed:
define models in a toucan2 way

dpsutton committed:
remove checksum: ANY from migrations

dpsutton committed:
api permissions 403 tests

dpsutton committed:
test how we fetch values

dpsutton commented on resources/migrations/000_migrations.yaml:
```diff
@@ -14615,6 +14615,198 @@ databaseChangeLog:
+              - column:
+                  name: generation
+                  type: int
+                  remarks: 'The current generation of indexed values. Values from previous generations can be deleted.'
```
I think the original wording is actually correct. It's not just submaximal. It's any row who's generation is not equal to the "current" generation as set on the `model_index` record.

metamben replied:
It would be quite confusing if "previous" generations could be greater than the current one (but NWM).

dpsutton commented on resources/migrations/000_migrations.yaml:
```diff
@@ -14615,6 +14615,198 @@ databaseChangeLog:
+                  remarks: 'The status of the index: initializing, indexed, error, overflow.'
+                  constraints:
+                    nullable: false
+
```
why yes it is.

dpsutton commented on src/metabase/models/model_index.clj:
```diff
@@ -0,0 +1,134 @@
+                     :query    {:source-table (format "card__%d" (:id model))
+                                :fields       [(:pk_ref model-index) (:value_ref model-index)]
+                                :order-by     [[:desc (:value_ref model-index)]]
+                                :limit        5001
```
yeah. the idea is we quit at 5000. and so if we have 5001 we know our state can be "overflow"

dpsutton commented on src/metabase/models/model_index.clj:
```diff
@@ -0,0 +1,134 @@
+                                :order-by     [[:desc (:value_ref model-index)]]
+                                :limit        5001
+                                :breakout   [(:pk_ref model-index) (:value_ref model-index)]}})
+                   :data :rows)]
```
explicitly don't want duplicates. But assuming the `pk_ref` is actually a primary key we gracefully handle duplicates but each has it's own id. The FE won't distinguish them but that's a bit of "holding it wrong" if the label for your index doesn't uniquely label the items.

dpsutton commented on src/metabase/models/model_index.clj:
```diff
@@ -0,0 +1,134 @@
+  (let [new-generation (inc (:generation model-index))]
+    ;; use upserts and delete ones with old generations
+    (t2/query {:insert-into   [:model_index_value]
+               :values        (into []
```
i think at the end of the day the `insert-into` will realize the entire collection as parameters. I don't think laziness or eductions help us unless we want to break this into several insert statements over some partitions of the whole.

dpsutton commented on src/metabase/models/model_index.clj:
```diff
@@ -0,0 +1,134 @@
+                                      :model_pk       id
+                                      :model_index_id (:id model-index)
+                                      :generation     (inc (:generation model-index))})
+                                   values)))
```
for mysql and postgres i do updates in place to (hopefully) prevent creating a lot of garbage in the table and reuse space. For h2 i don't worry about this and just delete them all and insert them all fresh. a bit of an optimization for the "non-production" h2 app db.

dpsutton commented on src/metabase/models/model_index.clj:
```diff
@@ -0,0 +1,134 @@
+                                     {:name           v
+                                      :model_pk       id
+                                      :model_index_id (:id model-index)
+                                      :generation     (inc (:generation model-index))})
```
effectively no, except that it follows the same invariants (all entries after indexed have the same generation as the model_index record). The idea is that rather than figuring out how to upsert in h2 just dump and recreate since it's not for bigger workloads.

dpsutton commented on src/metabase/models/model_index.clj:
```diff
@@ -0,0 +1,134 @@
+    ;; use upserts and delete ones with old generations
+    (t2/query {:insert-into             [:model_index_value]
+               :values                  (into []
+                                    (comp (filter (fn [[id v]] (and id v)))
```
indeed. fixed

dpsutton commented on src/metabase/models/model_index.clj:
```diff
@@ -0,0 +1,134 @@
+(defmethod add-values* :postgres
+  [model-index values]
+  (let [new-generation (inc (:generation model-index))]
+    ;; use upserts and delete ones with old generations
```
It's not just speed. I'm worried about constantly deleting rows and then adding the same ones back in. The worry is that this would accumulate lots of garbage in the table requiring a vacuum schedule. The hope is that the upserts can modify the existing row. If this proves not the case (either upserts create garbage or the vacuum worry is overblown) then yes mysql/postgres could just dump and recreate like h2 does.

metamben replied:
Are you sure that updates don't produce garbage? Updates in MVCC might be implemented as a delete+insert. The PostgreSQL docs contain this (vague) statement:
> VACUUM reclaims storage occupied by dead tuples. In normal PostgreSQL operation, tuples that are deleted or obsoleted by an update are not physically removed from their table; they remain present until a VACUUM is done.

dpsutton commented on src/metabase/models/model_index.clj:
```diff
@@ -0,0 +1,134 @@
+          (t2/update! ModelIndex (:id model-index)
+                      {:generation      generation'
+                       :state_change_at :%now
+                       :state           (if (> (count index-values) 5000) "overflow" "indexed")})))))
```
indeed. Should I move the `5000` to a constant or add a comment or leave it alone?

dpsutton committed:
remove empty line, add quotes to migration file

dpsutton committed:
cleanup from tamas's comments

- indention fix
- inline an inc
- add values in a tx (also catch errors and update model_index)
- clarify docstring on `should-deindex?`
- don't use declare, just define fn earlier

dpsutton commented on resources/migrations/000_migrations.yaml:
```diff
@@ -14615,6 +14615,198 @@ databaseChangeLog:
+                  constraints:
+                    nullable: false
+              - column:
+                  name: name
```
I haven't because I'm not sure that an index on it will improve the search we do. Do you think it would help?

metamben replied:
As you say, it depends on how these fields are searched. You mentioned full table scans, so I suppose we don't currently search in a way that would benefit from an index. But it's probably just a matter of effort.

dpsutton commented on resources/migrations/000_migrations.yaml:
```diff
@@ -14615,6 +14615,198 @@ databaseChangeLog:
+        - addUniqueConstraint:
+            tableName: model_index_value
+            constraintName: unique_model_index_value_index_id_model_pk
+            columnNames: model_pk, model_index_id
```
We need a uniqueness constraint and we need to have an index on `model_id`. I know in practice on postgres we can get both by putting the `model_index_id` first in the uniqueness constraint. Is this true across all of our app dbs (h2,pg,mysql) or are we leaning on an implementation detail that could change in future (or past?) versions?

metamben replied:
It's true for all out app DBs. (This is a property that falls out from BTREE indexes.)

dpsutton committed:
Merge branch 'master' into indexed-entities-schema

dpsutton committed:
update after merge

dpsutton committed:
update to new model name

dpsutton committed:
only test joins for drivers that support it

dpsutton committed:
don't test mongo, include scenario in test output

dpsutton committed:
sets use `disj` not `dissoc`

dpsutton committed:
handle oracle primary id type

`[94M "Awesome Bronze Plate"]`
(type 94M) -> java.math.BigDecimal

dpsutton committed:
remove magic value of 5000 for threshold

dpsutton committed:
Reorder unique constraint and remove extra index

previously had a unique constraint (model_pk, model_index_id) and an
index on model_index_id. If we reorder the unique constraint we get the
index on the model_index_id for free. So we'll take it.

dpsutton committed:
remove breakout.

without breakout

```sql
 SELECT "source"."id"    AS "ID",
       "source"."title" AS "TITLE"
FROM   (SELECT "PUBLIC"."products"."id"    AS "ID",
               "PUBLIC"."products"."title" AS "TITLE"
        FROM   "PUBLIC"."products") AS "source"
ORDER  BY "source"."title" DESC
```

with breakout

```sql
 SELECT "source"."id"    AS "ID",
       "source"."title" AS "TITLE"
FROM   (SELECT "PUBLIC"."products"."id"    AS "ID",
               "PUBLIC"."products"."title" AS "TITLE"
        FROM   "PUBLIC"."products") AS "source"
GROUP  BY "source"."id",
          "source"."title"
ORDER  BY "source"."title" DESC,
          "source"."id" ASC
```

dpsutton commented:
> Are you sure that updates don't produce garbage? Updates in MVCC might be implemented as a delete+insert. The PostgreSQL docs contain this (vague) statement:

    VACUUM reclaims storage occupied by dead tuples. In normal PostgreSQL operation, tuples that are deleted or obsoleted by an update are not physically removed from their table; they remain present until a VACUUM is done.

@metamben I'm not sure. But what I am sure of is that deleting them all and recreating them every day will create garbage. The hope is that the database can just update in place. If not, it's the same garbage as the naive approach but retains the possibility of not creating garbage.

Do you want to rip out the generation and just use the naive approach?

metamben commented:
> If not, it's the same garbage as the naive approach but retains the possibility of not creating garbage.

You can frame it that way. Another way of looking at it is "Why go into all this trouble of maintaining generations if it doesn't solve the problem it was meant to solve."

> Do you want to rip out the generation and just use the naive approach?

I would rip out generations if they don't significantly reduce the garbage generated. (In other words, I would test if they really help.)

iethree committed:
Merge branch 'master' into indexed-entities-schema

# Conflicts:
#	frontend/src/metabase-types/api/field.ts

dpsutton committed:
restore breakout

caused some tests with joins to fail

iethree committed:
update model-index api mock

dpsutton commented:
> dpsutton another option is querying the current set, leaving the intersection of old and new alone, and inserting/deleting the diff. that should produce minimal garbage, shouldn't it?

@metamben that was the first thought. Right now the set intersection for indexed values happens in the database. A true set intersection way requires loading both sets fully into memory and doing set operations on them. Right now our limit is 5000 values so memory isn't a huge concern. But doing it in memory could prove to be too expensive if this limit is raised or if we index many models.

I also prototyped a streaming version of the set intersection based on the indexed values. But it was so sensitive to string case that I ditched it. Basically it's a merge sort that only keeps a few values in memory at a time. This could probably work on the integer ids better as we would have an unambiguous order.

## Some investigation
I'm able to get tuple info with the following function:
```clojure
(defn tuple-stats
  []
  (let [x (-> (t2/query ["select n_live_tup, n_dead_tup, relname from pg_stat_all_tables where relname = 'model_index_value';"])
              (first))]
    {:live (:n_live_tup x)
     :dead (:n_dead_tup x)}))
```

And then we can go through 20 indexing jobs with the following:

```clojure
(reduce (fn [stats _]
          (model-index/add-values! model-index)
          (conj stats (doto (tuple-stats) tap>)))
        []
        (range 20))
```

#### Current generation

With the current implementation, we have the following sequence of garbage:
```
[{:live 200, :dead 0}
 {:live 200, :dead 0}
 {:live 200, :dead 0}
 {:live 200, :dead 400}
 {:live 200, :dead 400}
 {:live 200, :dead 400}
 {:live 200, :dead 400}
 {:live 200, :dead 770}
 {:live 200, :dead 770}
 {:live 200, :dead 787}
 {:live 200, :dead 787}
 {:live 200, :dead 825}
 {:live 200, :dead 825}
 {:live 200, :dead 825}
 {:live 200, :dead 825}
 {:live 200, :dead 818}
 {:live 200, :dead 818}
 {:live 200, :dead 818}
 {:live 200, :dead 818}
 {:live 200, :dead 854}]
```

#### Dump and recreate

```clojure
[{:live 200, :dead 0}
 {:live 200, :dead 200}
 {:live 200, :dead 200}
 {:live 200, :dead 600}
 {:live 200, :dead 600}
 {:live 200, :dead 600}
 {:live 200, :dead 800}
 {:live 200, :dead 800}
 {:live 200, :dead 800}
 {:live 200, :dead 800}
 {:live 200, :dead 800}
 {:live 200, :dead 1600}
 {:live 200, :dead 1600}
 {:live 200, :dead 1600}
 {:live 200, :dead 2200}
 {:live 200, :dead 2200}
 {:live 200, :dead 2200}
 {:live 200, :dead 2200}
 {:live 200, :dead 2200}
 {:live 200, :dead 3600}]
```

Feels safe to say that the generation count saves us quite a bit of garbage creation. If we want to do in-memory set operations we certainly could get this lower. What do you think?

metamben commented:
> I think it's possible to create a merge sort style set intersection that would produce far less garbage. But I think that implementation would be more complicated and sensitive than the current generation approach and leveraging the unique index on `(model_index_id, pk_id)` constraint that we currently use. What do you think?

If loading the 5000 existing pairs doesn't kill metabase, my preference would be to do that and make the minimal change in the DB. Limiting the size of the elements we load can help ensuring we can handle the data.
If we can't do that then I would go with the generation approach.
If later on we decide to significantly increase the 5000 limit (probably because the feature is important and people need it), then I would consider implementing the streaming comparison approach.

dpsutton committed:
Simpler method for indexing

Remove the `generation` part. We'll do the set operations in memory
rather than relying on db upserts. Now all app-dbs work the same and we
get all indexed values, diff against current values from the model, and
retract and add the appropriate values. "Updates" are counted as a
retraction and then addition rather than separately trying to update a
row.

Some garbage generation stats:

Getting garbage samples with

```clojure
(defn tuple-stats
  []
  (let [x (-> (t2/query ["select n_live_tup, n_dead_tup, relname from pg_stat_all_tables where relname = 'model_index_value';"])
              (first))]
    {:live (:n_live_tup x)
     :dead (:n_dead_tup x)}))
```

And using a simple loop to index the values repeatedly:

```clojure
(reduce (fn [stats _]
          (model-index/add-values! model-index)
          (conj stats (tuple-stats)))
        []
        (range 20))
```

With generation style:

```clojure
[{:live 200, :dead 0}
 {:live 200, :dead 0}
 {:live 200, :dead 0}
 {:live 200, :dead 400}
 {:live 200, :dead 400}
 {:live 200, :dead 400}
 {:live 200, :dead 400}
 {:live 200, :dead 770}
 {:live 200, :dead 770}
 {:live 200, :dead 787}
 {:live 200, :dead 787}
 {:live 200, :dead 825}
 {:live 200, :dead 825}
 {:live 200, :dead 825}
 {:live 200, :dead 825}
 {:live 200, :dead 818}
 {:live 200, :dead 818}
 {:live 200, :dead 818}
 {:live 200, :dead 818}
 {:live 200, :dead 854}]
```

With "dump and reload":

```clojure
[{:live 200, :dead 0}
 {:live 200, :dead 200}
 {:live 200, :dead 200}
 {:live 200, :dead 600}
 {:live 200, :dead 600}
 {:live 200, :dead 600}
 {:live 200, :dead 800}
 {:live 200, :dead 800}
 {:live 200, :dead 800}
 {:live 200, :dead 800}
 {:live 200, :dead 800}
 {:live 200, :dead 1600}
 {:live 200, :dead 1600}
 {:live 200, :dead 1600}
 {:live 200, :dead 2200}
 {:live 200, :dead 2200}
 {:live 200, :dead 2200}
 {:live 200, :dead 2200}
 {:live 200, :dead 2200}
 {:live 200, :dead 3600}]
```

And of course now that it's a no-op,

```clojure
[{:live 200, :dead 0}
 {:live 200, :dead 0}
 {:live 200, :dead 0}
 {:live 200, :dead 0}
 {:live 200, :dead 0}
 {:live 200, :dead 0}
 {:live 200, :dead 0}
 {:live 200, :dead 0}
 {:live 200, :dead 0}
 {:live 200, :dead 0}
 {:live 200, :dead 0}
 {:live 200, :dead 0}
 {:live 200, :dead 0}
 {:live 200, :dead 0}
 {:live 200, :dead 0}
 {:live 200, :dead 0}
 {:live 200, :dead 0}
 {:live 200, :dead 0}
 {:live 200, :dead 0}
 {:live 200, :dead 0}]
```

Since no db actions have taken when reindexing. We've traded memory set
operations for garbage in the database. Since we're capped at 5000 rows
it seems fine for now.

metamben commented on src/metabase/models/model_index.clj:
```diff
@@ -67,72 +67,50 @@
   [model-index]
-  (let [[error-message index-values] (fetch-values model-index)]
+  (let [[error-message values-to-index] (fetch-values model-index)
+        current-index-values               (into []
```
Why not load into a set right away? Now the order doesn't seem to be important.

metamben commented on src/metabase/models/model_index.clj:
```diff
@@ -67,72 +67,50 @@
-          (add-values* model-index (filter (fn [[_id value]] (some? value)) index-values))
+          (let [{:keys [additions deletions]} (find-changes {:current-index current-index-values
+                                                             :source-values values-to-index})]
+            (when-some [removal-ids (some->> deletions seq (map first))]
```
You could use `when` here too. That would be symmetrical with the insertions.

dpsutton committed:
add analogs to `isInteger` to constants.cljc and isa.cljc files

dpsutton committed:
clarity and do less work on set conversion

dpsutton committed:
exclude clojure.core/integer? from isa.cljc

dpsutton committed:
Rename `state_change_at` -> `indexed_at`

metamben reviewed (approved):
LGTM

dpsutton committed:
breakout brings along an order-by and fields clause
